Source/VCGeneration/VerificationConditionGenerator.cs

```diff
@@ -1445,7 +1430,7 @@ public static Counterexample AssertCmdToCloneCounterexample(VCGenOptions options
           {
             Contract.Assert(block != null);
             Contract.Assume(block.TransferCmd != null);
-            returnCmd = gotoCmdOrigins.GetValueOrDefault(block.TransferCmd);
+            returnCmd = block.TransferCmd.tok is GotoFromReturn gotoFromReturn ? gotoFromReturn.Origin : null;
```

Comment:
I now see how you encode meta-data in tokens.
I had launched a discussion about the design of having different tokens vs. flattened tokens here:
https://github.com/dafny-lang/dafny/issues/3354
It all boils down to deciding if you want to encode multiple information using wrapped tokens, where each token provides more information. If that is the case, then a flat token containing everything is more maintainable.
If the cases are truly disjoint, then a hierarchy of tokens is the way to go.
In the light of that, are your new tokens wrapper nestable or not?

Reply:
IMO the name `Token` should be reserved for lexical tokens, so data that only describes a source location and a word. However, I think for error reporting purposes, it is useful to track what the origin of nodes in the AST is, that origin might be `ParsedFromSource(IToken token) : IOrigin`, but it can also be something related to a program transformation. In this case, the origin of the goto jump is the desugaring transformation of a return jump.

Since I don't want to refactor tokens now, I'm abusing the IToken class as if it was an `IOrigin`.

Reply:
Agreed that `IToken` can be renamed to `IOrigin`, so that the token part is a sub-part of it, concerns only the lexical token, and not every metadata we don't know otherwise where to store.

In the meantime, if there is a token wrapper, I suggest there is at most one such wrapper that will contain any possible information about the origin. Is it what you did? Or are there at least two token wrapper classes?

Reply:
> Is it what you did? Or are there at least two token wrapper classes?

I think so, if I understand you correctly. Every AST node only has a single origin related field, whether that is a Token or a TokenWrapper.